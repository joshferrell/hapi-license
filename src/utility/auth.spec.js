import { dissoc } from 'ramda';
import createStrategy, { validateUser } from './auth';

describe('server authentication', () => {
    describe('validating a user object', () => {
        let decoded;
        let callback;
        beforeEach(() => {
            callback = jest.fn();
            decoded = {
                email: 'example@gmail.com',
                app_metadata: {
                    licenseCount: 2,
                    licenseTotal: 5
                }
            };
        });

        it('should reject if the app_meta data is empty', () => {
            decoded = dissoc('app_metadata', decoded);
            validateUser(decoded, null, callback);

            expect(callback).toHaveBeenCalled();
            expect(callback).toHaveBeenCalledTimes(1);
            expect(callback).toHaveBeenCalledWith(
                null,
                false
            );
        });

        it('should reject if the email is empty', () => {
            decoded = dissoc('email', decoded);
            validateUser(decoded, null, callback);

            expect(callback).toHaveBeenCalled();
            expect(callback).toHaveBeenCalledTimes(1);
            expect(callback).toHaveBeenCalledWith(
                null,
                false
            );
        });

        it('should reject if there is no licenseCount', () => {
            decoded = Object.assign({}, decoded, {
                app_metadata: {
                    licenseTotal: 0
                }
            });
            validateUser(decoded, null, callback);

            expect(callback).toHaveBeenCalled();
            expect(callback).toHaveBeenCalledTimes(1);
            expect(callback).toHaveBeenCalledWith(
                null,
                false
            );
        });

        it('should reject if there is no licenseTotal', () => {
            decoded = Object.assign({}, decoded, {
                app_metadata: {
                    licenseCount: 0
                }
            });
            validateUser(decoded, null, callback);

            expect(callback).toHaveBeenCalled();
            expect(callback).toHaveBeenCalledTimes(1);
            expect(callback).toHaveBeenCalledWith(
                null,
                false
            );
        });

        it('should allow if all fields are available', () => {
            validateUser(decoded, null, callback);

            expect(callback).toHaveBeenCalled();
            expect(callback).toHaveBeenCalledTimes(1);
            expect(callback).toHaveBeenCalledWith(
                null,
                true
            );
        });
    });

    describe('creating a server strategy', () => {
        let server;
        let hapiJwt2Key;
        beforeEach(() => {
            server = {
                auth: {
                    strategy: jest.fn()
                }
            };

            hapiJwt2Key = jest.fn(() => 'test');

            process.env.JWTKEY_JWKS_URI = 'test_jwks';
            process.env.VERIFY_OPTIONS_ISSUER = 'options_issue';
            process.env.VERIFY_OPTIONS_AUDIENCE = 'verify_audience';
        });

        it('should return a auth server strategy', () => {
            createStrategy(server, hapiJwt2Key);
            const strategy = server.auth.strategy.mock.calls[0];
            const jwtKey = hapiJwt2Key.mock.calls[0][0];

            expect(jwtKey).toMatchSnapshot();
            expect(strategy).toMatchSnapshot();
        });
    });
});
