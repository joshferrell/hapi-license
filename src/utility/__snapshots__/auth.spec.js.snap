// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server authentication creating a server strategy should return a auth server strategy 1`] = `
Object {
  "cache": true,
  "jwksRequestsPerMinute": 5,
  "jwksUri": "test_jwks",
  "rateLimit": true,
}
`;

exports[`server authentication creating a server strategy should return a auth server strategy 2`] = `
Array [
  "jwt",
  "jwt",
  "required",
  Object {
    "complete": true,
    "key": "test",
    "validateFunc": [Function],
    "verifyOptions": Object {
      "algorithms": Array [
        "RS256",
      ],
      "audience": "verify_audience",
      "issuer": "options_issue",
    },
  },
]
`;
